apps/wallet/src/background/account-sources/QredoAccountSource.ts

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+import { getAccountSources } from '.';
+import {
+	AccountSource,
+	type AccountSourceSerialized,
+	type AccountSourceSerializedUI,
+} from './AccountSource';
+import { accountSourcesEvents } from './events';
+import { backupDB, getDB } from '../db';
+import { type QredoConnectIdentity } from '../qredo/types';
+import { isSameQredoConnection } from '../qredo/utils';
+import { makeUniqueKey } from '../storage-utils';
+import { decrypt, encrypt } from '_src/shared/cryptography/keystore';
+import { QredoAPI } from '_src/shared/qredo-api';
+
+type DataDecryptedV0 = {
+	version: 0;
+	refreshToken: string;
+};
+
+type EphemeralData = {
+	refreshToken: string;
+	accessToken: string;
+};
+
+interface QredoAccountSourceSerialized extends AccountSourceSerialized, QredoConnectIdentity {
+	type: 'qredo';
+	encrypted: string;
+}
+
+interface QredoAccountSourceSerializedUI extends AccountSourceSerializedUI, QredoConnectIdentity {
+	type: 'qredo';
+}
+
+export class QredoAccountSource extends AccountSource<QredoAccountSourceSerialized, EphemeralData> {
+	constructor(id: string) {
+		super({ type: 'qredo', id });
+	}
+	static async createNew({
+		password,
+		apiUrl,
+		organization,
+		origin,
+		service,
+		refreshToken,
+	}: {
+		password: string;
+		apiUrl: string;
+		organization: string;
+		origin: string;
+		service: string;
+		refreshToken: string;
+	}) {
+		const decryptedData: DataDecryptedV0 = {
+			version: 0,
+			refreshToken,
+		};
+		const dataSerialized: QredoAccountSourceSerialized = {
+			id: makeUniqueKey(),
+			type: 'qredo',
+			apiUrl,
+			organization,
+			origin,
+			service,
+			encrypted: await encrypt(password, decryptedData),
+		};
+		const allAccountSources = await getAccountSources();
+		for (const anAccountSource of allAccountSources) {
+			if (
+				anAccountSource instanceof QredoAccountSource &&
+				isSameQredoConnection(
+					{
+						apiUrl: await anAccountSource.apiUrl,
+						organization: await anAccountSource.organization,
+						origin: await anAccountSource.origin,
+						service: await anAccountSource.service,
+					},
+					{ id: dataSerialized.id, apiUrl, organization, origin, service },
+				)
+			) {
+				throw new Error('Qredo account source already exists');
```

Comment:
Based on the way this is currently used throwing here I think is better. To make clear we can't store the new qredo source because it's a duplicate. There is check before calling this [here](https://github.com/MystenLabs/sui/blob/186ff82ba7f9e9d2b265df29c96c10b98468bffc/apps/wallet/src/background/qredo/index.ts#L235) so the error will probably never throw.